### DISCUSSION TITLE
Reference error when `x-bind` parameters within `x-for`

### DISCUSSION THREAD

**Discussion by GamesmenJordan** [alpinejs/alpine#2114] [7. Bugs (archived)]
### Discussed in https://github.com/alpinejs/alpine/discussions/2067

<div type='discussions-op-text'>

<sup>Originally posted by **GamesmenJordan** September 15, 2021</sup>
**Description**
When the parameters of an `x-for` loop are used within the context of an `x-bind` on elements within that loop, Alpine will throw a Reference Error for the parameter. This does not impact rendering each of the elements on the DOM and each element does appear to correctly have access to the parameter for other purposes (eg. `x-text` will display the string value of the parameter). But the Reference Error is uncaught and may break the execution of it or other scripts.

This issue does not happen on 2.x version of Alpine following the same setup.

**Example**
2.3.5 : https://jsfiddle.net/netjordanlee/2j71odwb/13/
3.3.3 : https://jsfiddle.net/netjordanlee/h632yxuz/9/

Both use the same Alpine data object:

```
function testComponent() {
	return {
        ready: false,
        myProperty: 'hello',
        myArray: [],
        init() {
            this.myArray = [
            {a:'1a', b:'1b'},
            {a:'2a', b:'2b'},
            {a:'3a', b:'3b'},
          ]

          this.ready = true;
        }
	}
}
```

And similar HTML (removed `x-init` because of the inferred call to `x-init` in 3.x)
```
<div x-data="testComponent()"
     x-init="init()">
     <span x-text="myProperty"></span>
  <template x-for="(element, index) in myArray" :key="element.a">
    <div :class="{ 'test' : (element.b === '2b') }">
      <span x-text="element.a"></span>
      <span x-text="element.b"></span>
    </div>
  </template>
</div>
```

In the console (use DevTools, not the JSFiddle Console) the following error appears in the 3.x version:
```
Uncaught (in promise) ReferenceError: element is not defined
    at eval (eval at generateFunctionFromString (cdn.js:364), <anonymous>:3:32)
    at cdn.js:375
    at tryCatch (cdn.js:399)
    at cdn.js:2241
    at reactiveEffect (cdn.js:901)
    at Object.effect2 [as effect] (cdn.js:876)
    at effect (cdn.js:44)
    at wrappedEffect (cdn.js:60)
    at Function.<anonymous> (cdn.js:2240)
    at flushHandlers (cdn.js:442)
```

Error points to the evaluation of `:class="{ 'test' : (element.b === '2b') }"` being the source of the error, presented as javascript below:
```
(async function anonymous(__self,scope
) {
with (scope) { __self.result = element.b }; __self.finished = true; return __self.result;
})
```</div>

**Comment by GamesmenJordan** [comment]
Confirmed this is still an issue with 3.3.5

**Reply by GamesmenJordan** [reply]
Confirmed this is still an issue with 3.4.4

**Reply by devalvarado200** [reply]
@GamesmenJordan Any solution to the issue? I'm working on a project and I'm getting the same error:

```
(async function anonymous(__self,scope
) {
with (scope) { __self.result = "wantSearch }; __self.finished = true; return __self.result;
})
```

**Comment by godismyjudge95** [comment]
I have also run into this issue. Can confirm that the reference error only occurs within the x-bind context.

Edit:  After further testing this appears to be happening only in conjunction with x-if filtering on x-for and then x-bind being used inside the x-for.  If I remove the x-if from GamesmenJordan's example it works without the reference error.  So maybe it tries to evaluate x-bind while the x-if is active and thus the element is not available yet?

**Comment by GamesmenJordan** [comment]
Someone posted another example (https://jsfiddle.net/kc5r2m0y/) in the discussions thread, it isn't restricted to x-if

**Reply by Yahav** [reply]
Yes it is, this fiddle is perfectly fine but notice that when you add x-if you'll get additional checkbox added:
https://jsfiddle.net/Lnocwe6u/

**Reply by GamesmenJordan** [reply]
No, it is not. See my example which also triggers the error without the use of x-if.

**Comment by Yahav** [comment]
Why was this closed?
its kind of a big deal..

**Reply by SimoTod** [reply]
This project doesn't use the issue tab: https://twitter.com/calebporzio/status/1430564014464159749?s=19

**Comment by Tim-Wils** [comment]
@SimoTod Need your insights, maybe I have found a fix for this :) 
Tried something in the x-if directive (hacked away in the javascript tab on line 2604) https://jsfiddle.net/eqm6cahy/2/

In short: The x-for directive fixes his own thing, so the if-directive only has to inject and init the template-contents if there's no el._x_lookup (x-for data) around.

**Reply by SimoTod** [reply]
I'll have a look tomorrow. Sorry for the late response, I was on holiday without my laptop so I couldn't really see your demo because jsfiddle is not responsive at all (codepen wins 😉).

**Reply by Tim-Wils** [reply]
Lol, I should make a codepen account asap 🤣 hope this could be a first try to the solution for this bug. The “if” should probably wrap around the mutateDom function though 😉

**Reply by SimoTod** [reply]
Your version removes the error but I believe that it will always print the x-for children even when x-if evaluates to false.
x-for is always evaluated before x-if (https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/directives.js#L159) so, as it stands today, there's really little to do unless we make x-for aware of the x-if tag. 
People should probably just accept that they can't put x-if and x-for on the same template tag, maybe an update to the documentation and a console warning.

**Reply by GamesmenJordan** [reply]
Please review the original discussion in https://github.com/alpinejs/alpine/discussions/2067 too, as this isn't an error that only effects the use of x-if (I believe this is actually a separate issue that has been conflated).

**Reply by SimoTod** [reply]
@GamesmenJordan i believe that is a separate issue (although similar outcome) triggered by the x-init, i was just giving a feedback for the x-for + x-if issue. 

For the other issue, someone needs to debug the source code and send a PR to fix it.

**Reply by Tim-Wils** [reply]
@SimoTod Thanks for the check! The fix indeed solves the errors, but unfortunately still prints the x-for. Maybe a custom directive of the x-for directive that's aware of the x-if directive would be a solution for this to work like Alpine 2.
I think how Alpine 2 worked vs Alpine 3 (the combined x-if x-for way) is what most people, including me are confused about.

**Comment by mayanksdudakiya** [comment]
@calebporzio With reference of https://github.com/alpinejs/alpine/issues/825 of Alpine issue. I am still facing this issue with the Alpine version `3.10.3`. So once the page successfully loaded first time then there was no any console error but when I tried to apply any filters and filter will reload the component at that time it shows the `Uncaught ReferenceError: is_theme_tab_opened_1 is not defined` error. Please have a look at the below screenshots for better understanding. 

Issue is happening with `x-bind:class` only


### Dependencies
![image](https://user-images.githubusercontent.com/34036151/192465333-b8424fa1-e525-4859-b7af-f5b9e45b9fd1.png)

### Code
`x-data` has already been defined but it still shows reference error in the browser console.

![image](https://user-images.githubusercontent.com/34036151/192465500-3e63a238-3333-4d07-9b86-20d179858983.png)

### Browser console error
![image](https://user-images.githubusercontent.com/34036151/192466045-25cdd8ad-0775-4a87-86b6-65316df9a55d.png)

### Inspected element with `x-data` already defined in <li>
![image](https://user-images.githubusercontent.com/34036151/192466299-e8bcceb9-ce50-48f8-b828-246d7b025776.png)

**Reply by SimoTod** [reply]
the error in the browser console mentions `is_theme_tab_opened_1` but the inspected html `shows opened_tab_1` did you post the right screenshot? If it's the right one, the error says that the variable is undefined which is true.

https://github.com/alpinejs/alpine/issues/825 was for alpine v2 I don't think it was ever ported to v3. Is there a reason why you can't declare your variable in x-data?

**Reply by mayanksdudakiya** [reply]
@SimoTod My bad. Here is the right screenshot. The variable is declared properly and on the first page load there is no any error but when I tried to apply search filter or any other filters, it start showing me the error in console.

![image](https://user-images.githubusercontent.com/34036151/192477328-372cd3d7-fd70-49d1-8c27-45e58096b428.png)

**Reply by SimoTod** [reply]
I can't see anything wrong so I suspect it's a bug with Alpine + Livewire. I don't use Livewire so, unfortunately, I can't help here.

**Reply by mayanksdudakiya** [reply]
@SimoTod No problem.
